uwnet/stochastic_parameterization/utils.py

```diff
@@ -0,0 +1,200 @@
+            true_nn_forcing[key] = (stop[key].values - start[key].values - (
+                self.time_step_seconds * forcing)) / self.time_step_days
+        return true_nn_forcing
+
+    def predict_for_time_idx(self, time_idx):
+        ds_filtered = self.ds.isel(time=time_idx)
+        to_predict = {}
+        for key_ in self.ds.data_vars:
+            if len(ds_filtered[key_].values.shape) == 2:
+                val = ds_filtered[key_].values[np.newaxis, :, :]
+            else:
+                val = ds_filtered[key_].values.astype(np.float64)
+            to_predict[key_] = torch.from_numpy(val)
+        pred = self.model(TensorDict(to_predict))
+        return {key: pred[key].detach().numpy() for key in pred}
+
+    def get_qt_ratios(self):
+        column_integrated_qt_residuals = np.ones_like(self.ds.Prec.values)
+        moistening_residuals = np.ones_like(self.ds.QT.values)
+        heating_residuals = np.ones_like(self.ds.SLI.values)
+        moistening_preds = np.ones_like(self.ds.QT.values)
+        heating_preds = np.ones_like(self.ds.SLI.values)
+        time_indices = list(range(len(self.ds.time) - 1))
+        idx = 0
+        for time_batch in np.array_split(time_indices, 10):
+            idx += 1
+            pred = self.predict_for_time_idx(time_batch)
+            true = self.get_true_nn_forcing_idx(time_batch)
+            q2_preds = np.ma.average(
+                pred['QT'],
+                axis=1,
+                weights=self.ds.layer_mass.values
+            )
+            q2_true = np.ma.average(
+                true['QT'],
+                axis=1,
+                weights=self.ds.layer_mass.values
+            )
+            column_integrated_qt_residuals[
+                time_batch, :, :] = q2_true - q2_preds
+            moistening_preds[time_batch, :, :, :] = pred['QT']
+            heating_preds[time_batch, :, :, :] = pred['SLI']
+            moistening_residuals[time_batch, :, :, :] = true['QT'] - pred['QT']
+            heating_residuals[time_batch, :, :, :] = true['SLI'] - pred['SLI']
+        self.heating_preds = heating_preds
+        self.moistening_preds = moistening_preds
+        self.column_integrated_qt_residuals = column_integrated_qt_residuals
+        self.moistening_residuals = moistening_residuals
+        self.heating_residuals = heating_residuals
+
+    def get_bin_membership(self):
+        self.get_qt_ratios()
+        if self.binning_method == 'precip':
+            bins = [
+                np.quantile(self.ds.Prec.values, quantile)
+                for quantile in self.binning_quantiles
+            ]
+            return np.digitize(
+                self.ds.Prec.values, bins, right=True)
+        elif self.binning_method == 'column_integrated_qt_residuals':
+            bins = [
+                np.quantile(self.column_integrated_qt_residuals, quantile)
+                for quantile in self.binning_quantiles
+            ]
+            return np.digitize(
```

Comment:
I don't think there is an `xarray` equivalent of digitize. In that case, I would define one:
```
def digitize(x, bins):
    # coerce to numpy array
    bins = np.asarray(bins)
    return xr.apply_ufunc(lambda x: np.digitize(x, bins), ds)
```